File: JavaProgram/bst_keysum.java
//Program to check if a BST contains two elements whose sum equals a given target.
import java.util.*;
//Node class
class TreeNode{
  int val;
  Node left;
  Node right;
//Constructor to create a new node with a given value 
  Node(int x){
    val=x;
    left=right=NULL;
  }
}
//Main Class
public class BST_KEYSUM{
  //Function to insert a value into the BST
  static Node insert(Node root,int val){
    if(root==NULL)
      //If tree is empty, create new node as root
      return new Node(val);
    if(val<root.val)
      //If value is amaller, insert into left subtree
      
    else
      //If value is larger or equal, insert into right subtree
      root.right=insert(root.right,val);
    return root;
  }

  //Inorder traversal: stores elements in sorted order into a list
  static void inorder(Node root,List<Integer> list){
    if(root==NULL) return;
    inorder(root.left,list);
    list.add(root.val);
    inorder(root.right,list);
  }
  //Function to check if there exists two elements whose sum equals k
  static boolean findTarget(Node root,int k){
    //Step1: store all BST in sorted order
    List<Integer> list=new ArrayList<>();
    inorder(root,list);
    //Step2: Use two-pointer technique on the sorted list
    int left=0;
    int right=list.size()-1;
    while(left<right){
      int sum=list.get(left)+list.get(right);
      if(sum==k)
        return true;
      else if(sum<k)
        left++;
      else
        right--;
    }
    return false;
  }
  //Main function
  public static void main(String[] args){
    Scanner sc=new Scanner(System.in);
    //Input number of nodes
    System.out.println("Enter number of nodes: ");
    int n=sc.nextInt();
    //Built the BST by inserting nodes one by one
    Node root=NULL;
    System.out.println("\n Enter node values: ");
    for(int i=0;i<n;i++){
      int val=sc.nextInt();
      root=insert(root,val);
    }
    //Input target sum
    System.out.print("\n Enter target sum: ");
    int k=sc.nextInt();
    //Call function and print result
    if(findTarget(root,k)){
      System.out.println("TRUE!");
      else
      System.out.println("FALSE!");
    sc.close();
  }
}

File: CProgram/BST_KEYSUM.c
//Given the root of a binary search tree and an integer k, return true if there exist two elements in the BST such that their sum is equal to k, or false otherwise.
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

struct Node {
    int val;
    struct Node* left,*right;
};

//create a node
struct Node* newNode(int value) {
    struct Node* node = (struct Node*) malloc (sizeof (struct Node));
    node->val=value;
    node -> left = node -> right = NULL;
    return node;
}
//insert a new node
struct Node* insert(struct Node* root,int value) {
    if(!root) return newNode(value);
    if(value < root->val)
    root->left = insert(root->left, value);
    else if(value > root->val)
       root->right = insert(root->right, value);
    return root;
}
//count the total number of nodes
int countNode(struct Node* root)
{
    if(!root)return 0;
    return 1 + countNode(root->left) + countNode(root->right);
}
//inorder insertion in an array
void inorder(struct Node* root, int arr[],int* index) {
    if(!root) return;
    inorder(root->left, arr, index);
    arr[(*index)++] = root->val;
    inorder(root->right, arr, index);

}
//main function to find whether the sum of two nodes is equal to the key provided
bool findTarget(struct Node* root,int k) {
    int n = countNode(root);
    int* arr = (int*)malloc(n * sizeof(int));
    int index = 0;
    inorder(root, arr, &index);
    int left=  0, right = n - 1;
    while(left < right) {
        int sum = arr[left] + arr[right];
        if(sum == k)
          return true;
        else if(sum < k)
          left++;
        else
          right++;
    }
    free(arr);
    return false;
}

int main() {
    struct Node* root = NULL;
    int n, i, val, k;
    printf("Enter the number of nodes: ");
    scanf("%d", &n);
    
    printf("\n Enter %d nodes of the BST tree: \n", n);
    for(i = 0; i < n; i++) {
      scanf("%4d", &val); 
      root = insert(root, val);
    }
    
    printf("\nEnter the target sum: ");
    scanf("%d", &k);
    if(findTarget(root, k))
     printf("\nTRUE!");
    else 
     printf("\nFALSE!");
    return 0;
}
